### PR TITLE
Remove getArgsNumAfterSegmentRuns

### PR DESCRIPTION
**Pull request by wujingyue**
Instead, check peak memory directly. The utilities will be used for verifying deallocation in host IR. 

https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html

### CODE REVIEW

File: tests/cpp/test_gpu3.cpp [hunk 1]
```diff
@@ -4859,12 +4859,15 @@ TEST_F(NVFuserTest, FusionSqueezeOnlyWelford_CUDA) {
 
   FusionExecutorCache executor_cache(std::move(fusion_ptr));
   auto cg_outputs = executor_cache.runFusionWithInputs({t0});
-  ASSERT_TRUE(at::allclose(
-      cg_outputs[0].as<at::Tensor>(), cg_outputs[3].as<at::Tensor>()));
-  ASSERT_TRUE(at::allclose(
-      cg_outputs[1].as<at::Tensor>(), cg_outputs[4].as<at::Tensor>()));
-  ASSERT_TRUE(at::allclose(
-      cg_outputs[2].as<at::Tensor>(), cg_outputs[5].as<at::Tensor>()));
+  ASSERT_TRUE(
```
Comment (wujingyue) [inline tests/cpp/test_gpu3.cpp:4862]: reformatted by lintrunner

**Comment by wujingyue** [comment]
!test

File: tests/cpp/utils.cpp [hunk 2]
```diff
@@ -43,6 +47,9 @@ void NVFuserTest::SetUp() {
 }
 
 void NVFuserTest::TearDown() {
+  at::cuda::clearCublasWorkspaces();
```
Comment (wujingyue) [inline tests/cpp/utils.cpp:50]: This is not strictly required -- I can also clearCublasWorkspaces only in RuntimeTest.ClearGmemBetweenSegments. But I think it's nice to verify no GPU memory leaks for all tests.

**Comment by wujingyue** [comment]
!test

**Comment by wujingyue** [comment]
!test

File: tests/cpp/test_runtime.cpp [hunk 3]
```diff
@@ -20,7 +21,11 @@ namespace nvfuser {
 
 using RuntimeTest = NVFuserTest;
 
-TEST_F(RuntimeTest, FusionClearGmemBetweenSegments_CUDA) {
+TEST_F(RuntimeTest, ClearGmemBetweenSegments) {
+  at::cuda::clearCublasWorkspaces();
+  releaseZeroedMemory();
+  ASSERT_EQ(memoryAllocated(0), 0) << "Memory leak detected";
```
Comment (wujingyue) [inline tests/cpp/test_runtime.cpp:27]: We can also do this in TearDown of all tests so we can verify none of them leak GPU memory. I'm happy to pursue that in a separate PR because the blast radius is larger.

**Comment by wujingyue** [comment]
!test

**Comment by wujingyue** [comment]
!test

File: tests/cpp/utils.cpp [hunk 4]
```diff
@@ -925,4 +929,39 @@ std::pair<at::Tensor, at::Tensor> createSdpaRngTensors() {
...
+int64_t memoryAllocated(const c10::DeviceIndex device) {
+  c10::cuda::CUDACachingAllocator::CUDAAllocator* allocator =
+      c10::cuda::CUDACachingAllocator::get();
+  NVF_CHECK(allocator != nullptr);
+
+  c10::CachingDeviceAllocator::DeviceStats device_stats =
+      allocator->getDeviceStats(device);
+
+  return device_stats.allocated_bytes
+      .at(static_cast<uint64_t>(c10::CachingAllocator::StatType::AGGREGATE))
```
Comment (jjsjann123) [inline tests/cpp/utils.cpp:963]:
```
#20 165.8 FAILED: CMakeFiles/nvfuser_multidevice_bench.dir/tests/cpp/utils.cpp.o 
#20 165.8 /opt/rh/gcc-toolset-12/root/usr/bin/c++ -DNVFUSER_DISTRIBUTED -DUSE_C10D_GLOO -DUSE_C10D_NCCL -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -I/workspace/Fuser -I/workspace/Fuser/csrc -I/workspace/Fuser/lib/dynamic_type/src -isystem /workspace/Fuser/third_party/benchmark/include -isystem /workspace/Fuser/third_party/flatbuffers/include -isystem /workspace/Fuser/third_party/googletest/googletest/include -isystem /workspace/Fuser/third_party/googletest/googletest -isystem /workspace/Fuser/third_party/gloo -isystem /opt/python/cp312-cp312/lib/python3.12/site-packages/torch/include -isystem /opt/python/cp312-cp312/lib/python3.12/site-packages/torch/include/torch/csrc/api/include -isystem /usr/local/cuda/include -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=1 -O3 -DNDEBUG -std=gnu++20 -fPIE -fvisibility=hidden -fvisibility-inlines-hidden -D_GLIBCXX_USE_CXX11_ABI=1 -MD -MT CMakeFiles/nvfuser_multidevice_bench.dir/tests/cpp/utils.cpp.o -MF CMakeFiles/nvfuser_multidevice_bench.dir/tests/cpp/utils.cpp.o.d -o CMakeFiles/nvfuser_multidevice_bench.dir/tests/cpp/utils.cpp.o -c /workspace/Fuser/tests/cpp/utils.cpp
#20 165.8 /workspace/Fuser/tests/cpp/utils.cpp: In function ‘int64_t nvfuser::maxMemoryAllocated(c10::DeviceIndex)’:
#20 165.8 /workspace/Fuser/tests/cpp/utils.cpp:950:38: error: ‘c10::CachingAllocator’ has not been declared
#20 165.8   950 |       .at(static_cast<uint64_t>(c10::CachingAllocator::StatType::AGGREGATE))
#20 165.8       |                                      ^~~~~~~~~~~~~~~~
#20 165.8 /workspace/Fuser/tests/cpp/utils.cpp: In function ‘int64_t nvfuser::memoryAllocated(c10::DeviceIndex)’:
#20 165.8 /workspace/Fuser/tests/cpp/utils.cpp:963:38: error: ‘c10::CachingAllocator’ has not been declared
#20 165.8   963 |       .at(static_cast<uint64_t>(c10::CachingAllocator::StatType::AGGREGATE))
```

I think you need to macro on torch version.

For 2.6 stable, this seems to be c10::CachingDeviceAllocator::StateType::AGGREGATE

Comment (jjsjann123) [reply tests/cpp/utils.cpp:963]: https://github.com/NVIDIA/Fuser/blob/f58617e8edfa13ad5e040046f3549d94402b3a45/csrc/utils.h#L37-L40

**Comment by wujingyue** [comment]
!build

**Review by jjsjann123 — APPROVED** [review]
